### PR TITLE
Issues/#93 sha3 pow

### PR DESCRIPTION
**Pull request by r-soltani**
This PR adds SHA3 capability to Block header and encoding, the change to SHA3 breaks the genesis block , therefore there is a commit in this PR that allows for recreation of the genesis block. Once a new genesis block is created, its hash can be added to the assertions found in chainparam.cpp

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -12,9 +13,23 @@
 
 #include <assert.h>
 #include <memory>
+#include <pow.h>
```
Comment (r-soltani) [inline src/chainparams.cpp]: This is to be surrounded by IFDEFs

File: src/hash.h [hunk 2]
```diff
@@ -15,6 +16,107 @@
 
 #include <vector>
 
+#ifndef BUILD_BTC
```
Comment (macterra) [inline src/hash.h]: Aren't these SHA3 functions already in hash.h?

File: src/hash.h [hunk 3]
```diff
@@ -15,6 +16,107 @@
...
+}
+
+/** Compute the 160-bit SHA-3 hash of a vector. */
+template<unsigned int N>
+inline uint160 SHA3Hash160(const prevector<N, unsigned char>& vch)
+{
+    return SHA3Hash160(vch.begin(), vch.end());
+}
+
+#endif
```
Comment (macterra) [inline src/hash.h]: This #endif should be at line 196, you appear to be missing SHA3SerializeHash() and supporting functions here.

File: src/primitives/block.cpp [hunk 4]
```diff
@@ -9,10 +10,15 @@
...
 #include <utilstrencodings.h>
 #include <crypto/common.h>
+#include <eqb/sha3/sha3.h>
 
 uint256 CBlockHeader::GetHash() const
 {
+#ifdef BUILD_BTC
     return SerializeHash(*this);
+#else  // BUILD_EQB
+    return SHA3::SerializeHash(*this);
```
Comment (macterra) [inline src/primitives/block.cpp]: Use SHA3SerializeHash from hash.h for consistency

File: src/hash.h (hunk 3 above)
Comment (r-soltani) [reply src/hash.h]: I will bring it into this PR.

File: src/primitives/block.cpp [hunk 5]
```diff
@@ -9,10 +10,17 @@
 #include <tinyformat.h>
 #include <utilstrencodings.h>
 #include <crypto/common.h>
+#ifndef BUILD_BTC
+#include <eqb/sha3/sha3.h>
```
Comment (macterra) [inline src/primitives/block.cpp]: No longer needed?

File: src/eqb/sha3/sha3.h [hunk 6]
```diff
@@ -35,6 +35,12 @@ class SHA3
         return *this;
     }
 
+    SHA3& write(char* pch, size_t size)
```
Comment (macterra) [inline src/eqb/sha3/sha3.h]: Changes to this file no longer needed?

**Comment by macterra** [comment]
Feel free to merge #143 to accept my proposed changes to this PR

File: src/primitives/block.cpp (hunk 5 above)
Comment (r-soltani) [reply src/primitives/block.cpp]: Indeed

**Review by macterra — APPROVED** [review]

**Comment by ealefad** [comment]
Node in this build doesn't keep the list of banned node in the `banlist.dat` file after the node restart. This is the reason for functional test `p2p_disconnect_ban.py` being failed. It looks like the node recreates a fresh (empty) `banlist.dat` every time it is being started. Way to manually reproduce:
1. `mkdir `/tmp/000`; `mkdir /tmp/000/node1`; `$ cd ...project/src`
2. start the node:
```
equibit-core/src$ ./equibitd -datadir=/tmp/000/node1 -rpcport=18332 -rpcuser=equibit -rpcpassword=equibit -regtest -server -keypool=1 -discover=0 -rest -logtimemicros -debug -debugexclude=libevent -debugexclude=leveldb -mocktime=0 -uacomment=testnode1
```
3. check that banned list is empty:
```
/equibit-core/src$ ./equibit-cli -datadir=/tmp/000/node1 -rpcport=18332 -rpcuser=equibit -rpcpassword=equibit listbanned
[
]
```
check the size of `banlist.dat`:
```
alex@ubuntu:/tmp/000/node1/regtest$ ls -la
total 48
drwx------ 5 alex alex 4096 Oct 12 16:45 .
drwxr-xr-x 3 alex alex 4096 Oct 12 16:45 ..
-rw------- 1 alex alex   37 Oct 12 16:45 banlist.dat
drwx------ 3 alex alex 4096 Oct 12 16:45 blocks
drwx------ 2 alex alex 4096 Oct 12 16:45 chainstate
-rw------- 1 alex alex 9022 Oct 12 16:45 debug.log
-rw------- 1 alex alex    6 Oct 12 16:45 equibitd.pid
-rw------- 1 alex alex    0 Oct 12 16:45 .lock
-rw------- 1 alex alex 4178 Oct 12 16:45 peers.dat
drwx------ 3 alex alex 4096 Oct 12 16:45 wallets

```
4. set a ban for 180 seconds:
```
/equibit-core/src$ ./equibit-cli -datadir=/tmp/000/node1 -rpcport=18332 -rpcuser=equibit -rpcpassword=equibit setban 192.0.111.1 add 180
```
verify the banlist:
```
/equibit-core/src$ ./equibit-cli -datadir=/tmp/000/node1 -rpcport=18332 -rpcuser=equibit -rpcpassword=equibit listbanned
[
  {
    "address": "192.0.111.1/32",
    "banned_until": 1539377361,
    "ban_created": 1539377181,
    "ban_reason": "manually added"
  }
]

```
check the size of
```
alex@ubuntu:/tmp/000/node1/regtest$ ls -la
total 48
drwx------ 5 alex alex 4096 Oct 12 16:46 .
drwxr-xr-x 3 alex alex 4096 Oct 12 16:45 ..
-rw------- 1 alex alex   91 Oct 12 16:46 banlist.dat
drwx------ 3 alex alex 4096 Oct 12 16:45 blocks
drwx------ 2 alex alex 4096 Oct 12 16:45 chainstate
-rw------- 1 alex alex 9757 Oct 12 16:46 debug.log
-rw------- 1 alex alex    6 Oct 12 16:45 equibitd.pid
-rw------- 1 alex alex    0 Oct 12 16:45 .lock
-rw------- 1 alex alex 4178 Oct 12 16:45 peers.dat
drwx------ 3 alex alex 4096 Oct 12 16:45 wallets

```

stop the nde
```
/equibit-core/src$ ./equibit-cli -datadir=/tmp/000/node1 -rpcport=18332 -rpcuser=equibit -rpcpassword=equibit stop
```

5. start the node again:
```
equibit-core/src$ ./equibitd -datadir=/tmp/000/node1 -rpcport=18332 -rpcuser=equibit -rpcpassword=equibit -regtest -server -keypool=1 -discover=0 -rest -logtimemicros -debug -debugexclude=libevent -debugexclude=leveldb -mocktime=0 -uacomment=testnode1
```

check list
```
/equibit-core/src$ ./equibit-cli -datadir=/tmp/000/node1 -rpcport=18332 -rpcuser=equibit -rpcpassword=equibit listbanned
[
]
```
check the file size (37 bytes again):
```
alex@ubuntu:/tmp/000/node1/regtest$ ls -la
total 308
drwx------ 5 alex alex   4096 Oct 12 16:48 .
drwxr-xr-x 3 alex alex   4096 Oct 12 16:45 ..
-rw------- 1 alex alex     37 Oct 12 16:48 banlist.dat
drwx------ 3 alex alex   4096 Oct 12 16:47 blocks
drwx------ 2 alex alex   4096 Oct 12 16:48 chainstate
-rw------- 1 alex alex  21256 Oct 12 16:48 debug.log
-rw------- 1 alex alex      6 Oct 12 16:48 equibitd.pid
-rw------- 1 alex alex 247985 Oct 12 16:47 fee_estimates.dat
-rw------- 1 alex alex      0 Oct 12 16:45 .lock
-rw------- 1 alex alex     17 Oct 12 16:47 mempool.dat
-rw------- 1 alex alex   4178 Oct 12 16:48 peers.dat
drwx------ 3 alex alex   4096 Oct 12 16:48 wallets

```

The node from `dev` doesn't rewrite the `banlis.dat` and has a list of banned nodes after the node restart, e.g.:
```
alex@ubuntu:/tmp/000/node1/regtest$ ls -la
total 48
drwx------ 5 alex alex 4096 Oct 12 16:46 .
drwxr-xr-x 3 alex alex 4096 Oct 12 16:45 ..
-rw------- 1 alex alex   91 Oct 12 16:46 banlist.dat
drwx------ 3 alex alex 4096 Oct 12 16:45 blocks
drwx------ 2 alex alex 4096 Oct 12 16:45 chainstate
-rw------- 1 alex alex 9757 Oct 12 16:46 debug.log
-rw------- 1 alex alex    6 Oct 12 16:45 equibitd.pid
-rw------- 1 alex alex    0 Oct 12 16:45 .lock
-rw------- 1 alex alex 4178 Oct 12 16:45 peers.dat
drwx------ 3 alex alex 4096 Oct 12 16:45 wallets
```

**Review by ealefad — CHANGES_REQUESTED** [review]
Please see my comment https://github.com/Equibit/equibit-core/pull/136#issuecomment-429454131

**Review by ealefad — APPROVED** [review]

**Comment by ealefad** [comment]
Functional tests `p2p_disconnect.ban` are temporarily disabled. New issue for this: https://app.zenhub.com/workspace/o/equibit/equibit-core/issues/149

**Pull request merged**